iep-004/README.adoc

```diff
@@ -0,0 +1,260 @@
+instances in order to provide an ideal balance of cost and performance. For
+more, see <<Costs>> below.
+
+
+*Kubernetes master*
+
+A cluster would run a single D3v2 instance as the master node
+
+
+*Kubernetes agents*
+
+A cluster would run agents using a
+link:https://azure.microsoft.com/en-us/services/virtual-machine-scale-sets/[Scale Set]
+of D2v2 instances, with a minimum of three agents running at all times.
+
+
+
+[NOTE]
+====
+Following
+link:https://github.com/jenkins-infra/iep/tree/master/iep-003[IEP-3]
+the <<reference-implementation>> uses Terraform to describe the infrastructure
+necessary to support this specification.
+====
+
+
+=== Monitoring
+
+As the Jenkins project already uses
+link:http://datadoghq.com[Datadog]
+for monitoring our production infrastructure, the Kubernetes cluster must
+integrate appropriately.
+
+This will be accomplished using the
+link:http://docs.datadoghq.com/integrations/kubernetes/[Datadog/Kubernetes integration]
+maintained by Datadog themselves.
+
+The integration is will provide metrics for both:
+
+* Kubernetes cluster health/status
+* Container health/status running atop the cluster
+
+The metrics required are not specified here under the assumption that all
+metrics appropriate for ensuring stable service levels of project
+infrastructure will be collected.
+
+=== Logging
+
+[NOTE]
+====
+This section is still a work in progress, awaiting information to be provided
+by
+link:https://github.com/olblak[Olivier Vernin]
+====
+
+Centralized logging for applications hosted within Kubernetes will be provided
+a combination of containers in the cluster running
+link:https://en.wikipedia.org/wiki/Fluentd[Fluentd]
+and the Microsoft
+link:http://www.microsoft.com/en-us/cloud-platform/operations-management-suite[Operations Management Suite]
+(OMS).
+
+.Data flow for k8s logs
+[source]
+----
++-------------------------+
+|  Kubernetes             |
+| |          ^            |
+| | +-----+  | Logs to k8s|
+| | | App +--`
+| | +-----+               |          +-----------------------+
+| | +---------+   Publish to blobs   |                       |
+| `>| Fluentd +--------------------->|   Azure Blob Storage  |
+|   +---------+           |          |                       |
+|   +-----+               |          +-----------+-----------+
+|   | App |               |                      |
+|   +-----+               |                      | Pull from blobs as data source
++-------------------------+                      v
+                                  +-------------------------------+
+                                  |  Operations Management Suite  |
+                                  +-------------------------------+
+----
+
+
+
+=== Deployment/Orchestration
+
+[NOTE]
+====
+This section is still a work in progress, awaiting and prototyping by
+link:https://github.com/olblak[Olivier Vernin]
+====
+
+== Motivation
+
+The motivation for centralizing container hosting is fairly
```

Comment:
It will also be easier to replicate containers hosting for different environments.